lib/cancancan/neo4j/cypher_constructor.rb

```diff
@@ -0,0 +1,83 @@
+require 'cancancan/neo4j/cypher_constructor_helper'
+
+module CanCanCan
+  module Neo4j
+    # Constructs cypher query from rule cypher options
+    class CypherConstructor
+      attr_reader :query
+
+      def initialize(rule_cyphers)
+        @model_class = rule_cyphers.first.options[:model_class]
+        reset_variables
+        construct_cypher(rule_cyphers)
+      end
+
+      def reset_variables
+        @query = @model_class.new_query
+        @current_collection = nil
+      end
+
+      def construct_cypher(rule_cyphers)
+        rule_cyphers.each do |rule_cypher|
+          rule = rule_cypher.options[:rule]
+          reset_variables if rule.conditions.blank?
+          if rule.base_behavior
+            construct_can_cypher(rule_cypher)
+          else
+            construct_cannot_cypher(rule_cypher)
+          end
+        end
+        unwind_query_with_distinct
+      end
+
+      def unwind_query_with_distinct
+        var = CanCanCan::Neo4j::CypherConstructorHelper.var_name(@model_class)
+        @query = unwind_qeury("#{var}_can")
+                 .with("DISTINCT #{var}_can as #{var}")
+      end
+
+      def unwind_qeury(var_name)
+        @query = @query.unwind("#{@current_collection} as #{var_name}")
+      end
+
+      def construct_can_cypher(rule_cypher)
+        with_clause = with_clause_for_rule(rule_cypher, true)
+        @query = @query.optional_match(rule_cypher.path)
+                       .where(rule_cypher.rule_conditions)
+                       .with(with_clause)
+      end
+
+      def with_clause_for_rule(rule_cypher, can_rule)
+        var = rule_cypher.options[:var_label]
+        with = "collect(DISTINCT #{var}) as #{var}_col"
+        if can_rule && @current_collection
+          with = @current_collection + ' + ' + with
```

Comment:
looking at the usage of `+` it occurred to me why we don't have codeclimate analysis available?
generally string concatenation is better done with interpolation or `join`.

Reply:
We have codeclimate setup https://codeclimate.com/github/CanCanCommunity/cancancan-neo4j